contracts/compounder/GenericVault.sol

```diff
@@ -0,0 +1,335 @@
+        for (uint256 i = 0; i < len; i++) {
+            emit ExtraRewardCleared(extraRewards[i]);
+        }
+        delete extraRewards;
+    }
+
+    /// @notice Query the amount currently staked
+    /// @return total - the total amount of tokens staked
+    function totalUnderlying() public view returns (uint256 total) {
+        return IStrategy(strategy).totalUnderlying();
+    }
+
+    /// @notice Returns the amount of underlying a user can claim
+    /// @param user - address whose claimable amount to query
+    /// @return amount - claimable amount
+    /// @dev Does not account for penalties and fees
+    function balanceOfUnderlying(address user) external view returns (uint256 amount) {
+        require(totalSupply() > 0, "No users");
+        return ((balanceOf(user) * totalUnderlying()) / totalSupply());
+    }
+
+    /// @notice Deposit user funds in the autocompounder and mints tokens
+    /// representing user's share of the pool in exchange
+    /// @param _amount - the amount of underlying to deposit
+    /// @return _shares - the amount of shares issued
+    function deposit(uint256 _amount, address _receiver)
+        public
+        notToZeroAddress(_receiver)
+        nonReentrant
+        returns (uint256 _shares)
+    {
+        require(_amount > 0, "Deposit too small");
+
+        // Stake into extra rewards before we update the users
+        // balancers and update totalSupply/totalUnderlying
+        for (uint256 i = 0; i < extraRewards.length; i++) {
+            IBasicRewards(extraRewards[i]).stake(_receiver, _amount);
+        }
+
+        uint256 _before = totalUnderlying();
+        IERC20(underlying).safeTransferFrom(msg.sender, strategy, _amount);
+        IStrategy(strategy).stake(_amount);
+
+        // Issues shares in proportion of deposit to pool amount
+        uint256 shares = 0;
+        if (totalSupply() == 0) {
+            shares = _amount;
+        } else {
+            shares = (_amount * totalSupply()) / _before;
+        }
+
+        _mint(_receiver, shares);
+        emit Deposit(msg.sender, _receiver, _amount, shares);
+        return shares;
+    }
+
+    /// @notice Unstake underlying in proportion to the amount of shares sent
+    /// @param _shares - the number of shares sent
+    /// @return _withdrawable - the withdrawable underlying amount
+    function _withdraw(address _from, uint256 _shares) internal returns (uint256 _withdrawable) {
+        require(totalSupply() > 0);
+        // Computes the amount withdrawable based on the number of shares sent
+        uint256 amount = (_shares * totalUnderlying()) / totalSupply();
+        // Burn the shares before retrieving tokens
+        _burn(_from, _shares);
+        // If user is last to withdraw, harvest before exit
+        if (totalSupply() == 0) {
+            harvest();
+            IStrategy(strategy).withdraw(totalUnderlying());
+            _withdrawable = IERC20(underlying).balanceOf(address(this));
```

Comment:
Edge case: last user in the vault takes everything sent to the vault by accident during the vaults lifespan ("dust"), as well as there rightful holdings. Can we deal with vault "dust" in any way (ie forward it to the strategy and then harvest, or alternatively, subtract the start balance from the end balance and send the diff to treasury?

Reply:
I think this is fine. We want to try and minimize the diff on the llama airforce contracts we forked and so think we can leave as is.